pkg/signup/service/signup_service.go

```diff
@@ -589,33 +595,49 @@ func (s *ServiceImpl) DoGetUserSignupFromIdentifier(provider ResourceProvider, u
 
 // UpdateUserSignup is used to update the provided UserSignup resource, and returning the updated resource
 func (s *ServiceImpl) UpdateUserSignup(userSignup *toolchainv1alpha1.UserSignup) (*toolchainv1alpha1.UserSignup, error) {
-	userSignup, err := s.CRTClient().V1Alpha1().UserSignups().Update(userSignup)
-	if err != nil {
+	if err := s.Update(gocontext.TODO(), userSignup); err != nil {
 		return nil, err
 	}
 
 	return userSignup, nil
 }
 
+var (
+	md5Matcher = regexp.MustCompile("(?i)[a-f0-9]{32}$")
+)
+
 // PhoneNumberAlreadyInUse checks if the phone number has been banned. If so, return
-// an internal server error. If not, check if an active (non-deactivated) UserSignup with a different userID and username
+// an internal server error. If not, check if an approved UserSignup with a different userID and username
 // and email address exists. If so, return an internal server error. Otherwise, return without error.
 // Either the actual phone number, or the md5 hash of the phone number may be provided here.
 func (s *ServiceImpl) PhoneNumberAlreadyInUse(userID, username, phoneNumberOrHash string) error {
-	bannedUserList, err := s.CRTClient().V1Alpha1().BannedUsers().ListByPhoneNumberOrHash(phoneNumberOrHash)
-	if err != nil {
+	labelValue := hash.EncodeString(phoneNumberOrHash)
+	if md5Matcher.Match([]byte(phoneNumberOrHash)) {
+		labelValue = phoneNumberOrHash
+	}
+
+	bannedUserList := &toolchainv1alpha1.BannedUserList{}
+	if err := s.List(gocontext.TODO(), bannedUserList, client.InNamespace(s.Namespace),
+		client.MatchingLabels{toolchainv1alpha1.BannedUserPhoneNumberHashLabelKey: labelValue}); err != nil {
 		return errors.NewInternalError(err, "failed listing banned users")
 	}
+
 	if len(bannedUserList.Items) > 0 {
 		return errors.NewForbiddenError("cannot re-register with phone number", "phone number already in use")
 	}
 
-	userSignups, err := s.CRTClient().V1Alpha1().UserSignups().ListActiveSignupsByPhoneNumberOrHash(phoneNumberOrHash)
-	if err != nil {
+	labelSelector := client.MatchingLabels{
+		toolchainv1alpha1.UserSignupStateLabelKey:           toolchainv1alpha1.UserSignupStateLabelValueApproved,
+		toolchainv1alpha1.BannedUserPhoneNumberHashLabelKey: labelValue,
+	}
+	userSignups := &toolchainv1alpha1.UserSignupList{}
+	if err := s.List(gocontext.TODO(), userSignups, client.InNamespace(s.Namespace), labelSelector); err != nil {
 		return errors.NewInternalError(err, "failed listing userSignups")
 	}
-	for _, signup := range userSignups {
-		if signup.Spec.IdentityClaims.Sub != userID && signup.Spec.IdentityClaims.PreferredUsername != username && !states.Deactivated(signup) { // nolint:gosec
+
+	for _, signup := range userSignups.Items {
+		userSignup := signup
```

Comment:
is it needed? IIRC, `signup` is not a pointer

Reply:
no, but I create a pointer to the `userSignup` in the next line and since we still use go 1.20 we have to do this re-assignment

Reply:
added a comment 8df7f52

Reply:
great, thanks!